### ISSUE TITLE
An invalid response was received from the upstream server.

### ISSUE THREAD

**Issue by hongkang** [Mashape/kong#2360]
I use Kong 0.10.1.
It happens occasionally when in the first request.
And the second request is always normal.

**Comment by bungle** [comment]
@hongkang, can you provide a little bit more information. E.g. `headers`and `bodies` of first request vs. the second?

**Comment by saamalik** [comment]
We're also getting intermittent `invalid response` was received from upstream server. Using `FROM mashape/kong:0.10.0rc3`. 

I'll try to capture the kong logs next time this happens.

**Comment by amarjeet27** [comment]
Reverting back to `kong:0.9.9` seems to help.

**Comment by saamalik** [comment]
Thanks, @amarjeet27 ; `kong:0.9.9` does seem to help. I was able to capture the intermittent messages in `error.log`:


```
2017/04/16 22:54:04 [error] 94#0: *99116 recv() failed (104: Connection reset by peer) while reading response header from upstream, client: 10.0.0.7, server: kong, request: "GET /api/v1/sites/58f28f767800007800441078/aci/domains?type=vmm HTTP/1.0", upstream: "http://10.0.0.14:9010/api/v1/sites/58f28f767800007800441078/aci/domains?type=vmm", host: "kong:8000", referrer: "http://localhost/schemas/edit/58f3f6008600009100c83066"
2017/04/16 22:54:04 [error] 94#0: *99116 recv() failed (104: Connection reset by peer) while reading response header from upstream, client: 10.0.0.7, server: kong, request: "GET /api/v1/sites/58f28f767800007800441078/aci/domains?type=vmm HTTP/1.0", upstream: "http://10.0.0.14:9010/api/v1/sites/58f28f767800007800441078/aci/domains?type=vmm", host: "kong:8000", referrer: "http://localhost/schemas/edit/58f3f6008600009100c83066"
2017/04/16 22:54:26 [error] 94#0: *99212 recv() failed (104: Connection reset by peer) while reading response header from upstream, client: 10.0.0.7, server: kong, request: "GET /api/v1/schemas HTTP/1.0", upstream: "http://10.0.0.16:9020/api/v1/schemas", host: "kong:8000", referrer: "http://localhost/schemas"
2017/04/16 23:01:32 [error] 95#0: *101074 recv() failed (104: Connection reset by peer) while reading response header from upstream, client: 10.0.0.7, server: kong, request: "GET /api/v1/aci/sites/58f28f767800007800441078/nodes?type=leaf HTTP/1.0", upstream: "http://10.0.0.14:9010/api/v1/aci/sites/58f28f767800007800441078/nodes?type=leaf", host: "kong:8000", referrer: "http://localhost/schemas/edit/58f3f6008600009100c83066"
2017/04/16 23:03:34 [error] 95#0: *101598 recv() failed (104: Connection reset by peer) while reading response header from upstream, client: 10.0.0.7, server: kong, request: "PUT /api/v1/schemas/58f3f6008600009100c83066 HTTP/1.0", upstream: "http://10.0.0.16:9020/api/v1/schemas/58f3f6008600009100c83066", host: "kong:8000", referrer: "http://localhost/schemas/edit/58f3f6008600009100c83066"
2017/04/16 23:23:14 [error] 95#0: *106684 recv() failed (104: Connection reset by peer) while reading response header from upstream, client: 10.255.0.3, server: kong, request: "GET /api/v1/execute/schema/58f3f6008600009100c83066/template/template1 HTTP/1.1", upstream: "http://10.0.0.2:9030/api/v1/execute/schema/58f3f6008600009100c83066/template/template1", host: "localhost:8000"
2017/04/16 23:27:45 [error] 96#0: *107838 recv() failed (104: Connection reset by peer) while reading response header from upstream, client: 10.255.0.3, server: kong, request: "PUT /api/v1/schemas/58f3f6008600009100c83066 HTTP/1.1", upstream: "http://10.0.0.16:9020/api/v1/schemas/58f3f6008600009100c83066", host: "localhost:8000"
2017/04/16 23:31:28 [error] 94#0: *108807 recv() failed (104: Connection reset by peer) while reading response header from upstream, client: 10.255.0.3, server: kong, request: "GET /api/v1/schemas HTTP/1.1", upstream: "http://10.0.0.16:9020/api/v1/schemas", host: "localhost:
```

**Comment by p0pr0ck5** [comment]
The logs above only indicate that the upstream closed the connection before it finished sending data to the upstream.

BTW there are many people involved in this thread, and we will need some clear, concise detail to make sense of it all :)

**Comment by saamalik** [comment]
@p0pr0ck5 I'll update to the latest version of Kong again to reproduce these errors, but what more logs would you like to see?

**Comment by p0pr0ck5** [comment]
This log entry:

`*108807 recv() failed (104: Connection reset by peer) while reading response header from upstream`

Means that the connection was reset by the upstream, not by Kong. Are there any other notable entries in the error log? Any errors in the upstream nodes?

**Comment by bolv88** [comment]
It also happens to me ! 
My log level is notice , there is not other log!

**Comment by saamalik** [comment]
@bolv88 try downgrading to `kong:0.9.9` to see if the issue goes away.

**Comment by p0pr0ck5** [comment]
@saamalik any update based on the last response:

> This log entry:
>
> *108807 recv() failed (104: Connection reset by peer) while reading response header from upstream
>
> Means that the connection was reset by the upstream, not by Kong. Are there any other notable entries in the error log? Any errors in the upstream nodes?

**Comment by saamalik** [comment]
@p0pr0ck5 we upgraded to `0.10.3` and sadly still intermittently see the same problem. Like @hongkang mentioned, the failure is only seen on generally the first request after some idle-time. If I was to take a _wild_ guess, I would say something to do with `keep-alive` which is enabled in the the `0.10.x` series. 

Backend servers are web _playframework_ http servers. The request is not even initiated (nothing in the access logs of the backend server). Reverting back to `0.9.9` and the issue again is not reproducible. 

Any other ideas how to debug?

**Comment by saamalik** [comment]
Just for reference:
Kong `0.10.3` was deployed using docker. 
Database: Postgres
Upstream services: Playframework (2.5.11) (in Docker Swarm)

**Comment by saamalik** [comment]
Ok last point for reference. Playframework web server is netty based. I enabled the netty debug logs which shows all relevant information regarding the request-response (headers, body, response, etc).

Now in the happy paths, requests and responses are printed normally. To reproduce the Kong issue, just stay idle for ~10 min; and then trigger an API call. Everytime the first connection after a long bout of idle results in the "connection reset". And unfortunately, nothing is printed on the server side--zilch, almost as if the request never even made it here. Second request _always_ goes through.

**Comment by p0pr0ck5** [comment]
@saamalik thanks for the report back.

It sounds like your upstream is not respecting keepalive properly, and closing the connection without notifying downstream (Kong). What does tcpdump on the playframework hosts tell you when Kong sends  a request after a long idle period? My guess is the TCP connection is no longer held in an ESTABLISHED state upstream, which is something Kong cannot help with. And if it was never told to close the keepalive connection, it will assume the same TCP connection is usable, and not initiate a new handshake, meaning your upstream servers see packets for a connection they aren't tracking, and send back an RST.

So tl;dr, check your upstream keepalive timeout/network settings :)

**Comment by p0pr0ck5** [comment]
Closing this as stale, following no response (and a thumbs-up from @saamalik :) ). If there are any further issues to report with Kong, please let us know. Thank you!

**Comment by saamalik** [comment]
Sorry for the delay. I'll try to implement TCP SO_KEEPALIVE in Netty server
tomorrow.
